### PR TITLE
Delay branch updates from workshop

### PR DESCRIPTION
**Pull request by stavness**
- moved testDelay.cpp from "OpenSim/Common/Test" to  "OpenSim/Simulation/Test" (because it requires osimSimulation.h and osimActuators.h)
- moved Delay.h from OpenSim/Common to OpenSim/Simulation/Model
- renamed constructInfrastructure to constructProperties; removed override keyword
- include only required headers
- in extendConnectToModel added if statements to check for a connection before connecting inputs, because method is called multiple times
- fixed testNegativeDelayDurationException — SimTK_TEST_MUST_THROW_EXC for model.addController() call
- in DiscontinuousInputTesting: added set name method; changed three Delay components to be member subcomponents; added memberSubcomponentIndex to class; added new calls to get Delay component output value

### CODE REVIEW

File: OpenSim/Simulation/Test/testDelay.cpp [hunk 1]
```diff
@@ -603,14 +632,19 @@ void testDiscontinuousInput() {
     // The "step" output switches at 0.5 seconds, but the delay is 0.2 seconds,
     // so the delay output should still be 1.5.
     //std::cout << comp->stepDelay.getValue(ts.getState()) << " " << std::endl;
-    SimTK_TEST_EQ(comp->stepDelay.getValue(ts.getState()), 1.5);
-    //std::cout << comp->discontDelay.getValue(ts.getState()) << " " << std::endl;
-    SimTK_TEST_EQ(comp->discontDelay.getValue(ts.getState()), 0.46);
 
+    const Delay& stepDelay = comp->getComponent<Delay>("/discontinuousInputTesting/stepDelay");
+    const Delay& discontDelay = comp->getComponent<Delay>("/discontinuousInputTesting/discontDelay");
```
Comment (chrisdembia) [inline OpenSim/Simulation/Test/testDelay.cpp]: The current way to fish out a member subcomponent is shown [here](https://github.com/opensim-org/opensim-core/blob/de24648dbc6c1342d74bf7ba8b792fd4c08edbc6/OpenSim/Actuators/Thelen2003Muscle.cpp#L103).

**Comment by chrisdembia** [comment]
> in extendConnectToModel added if statements to check for a connection before connecting inputs, because method is called multiple times

I think this is fine but I am curious. Is this for performance? Was reconnecting slow?

Thanks for updating this, @stavness. I'll merge this now.

**Comment by chrisdembia** [comment]
Now #398 is updated, and further work on the Delay component can happen there. Let me know when you think #398 is ready to be reviewed, or if you have things planned or other things that you think should be done first.

**Comment by mshabani** [comment]
@chrisdembia  
"in extendConnectToModel added if statements to check for a connection before connecting inputs, because method is called multiple times"
I think this is fine but I am curious. Is this for performance? Was reconnecting slow?

Removing these if statements would result in an exception complaining about connecting input multiple time. That's because "extendConnectToModel " is called several times internally and every time it tries to connect all inputs. first time it's ok because those are not connected. in the second call exception happens.
Sorry I don't have access to reply to your comment so I wrote a new one.

**Comment by chrisdembia** [comment]
Thank you for explaining.

**Pull request merged**